server/src/main/java/org/opensearch/plugins/ResourceAccessControlPlugin.java

```diff
@@ -0,0 +1,99 @@
+/*
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * The OpenSearch Contributors require contributions made to
+ * this file be licensed under the Apache-2.0 license or a
+ * compatible open source license.
+ */
+
+package org.opensearch.plugins;
+
+import org.opensearch.accesscontrol.resources.RecipientType;
+import org.opensearch.accesscontrol.resources.Resource;
+import org.opensearch.accesscontrol.resources.ResourceSharing;
+import org.opensearch.accesscontrol.resources.ShareWith;
+
+import java.util.Map;
+import java.util.Set;
+
+/**
+ * This plugin allows to control access to resources. It is used by the ResourcePlugins to check whether a user has access to a resource defined by that plugin.
+ * It also defines java APIs to list, share or revoke resources with other users.
+ * User information will be fetched from the ThreadContext.
+ *
+ * @opensearch.experimental
+ */
+public interface ResourceAccessControlPlugin {
+
+    /**
+     * Returns all accessible resources for current user for a given plugin index.
+     * @param resourceIndex index where the resource exists
+     * @param clazz class of the resource. Required to parse the resource object retrieved from resourceIndex. Must be a type of {@link Resource}
+     * @return set of {@link ResourceSharing} items accessible by current user.
+     */
+    default <T extends Resource> Set<T> getAccessibleResourcesForCurrentUser(String resourceIndex, Class<T> clazz) {
+        return Set.of();
+    }
+
+    /**
+     * Checks whether current user has permission to given resource.
+     *
+     * @param resourceId the resource on which access is to be checked
+     * @param resourceIndex where the resource exists
+     * @param scope the scope being requested
+     * @return true if current user has access, false otherwise
+     */
+    default boolean hasPermission(String resourceId, String resourceIndex, String scope) {
```

Comment:
What is a scope exactly? Is it the same as security's notion of an action group or something different? How is it defined and why is it a String?

If `resourceIndex` is passed in here then is it assumed that resources are stored in an index? Should this be abstracted to not necessarily assume that resources are stored in an index and leave the implementation up to the plugin that provides access control for resources?

Reply:
Would this always be able to computed synchronously? If it can't be synchronous then does this need to accept an ActionListener?